runtime/permissions.rs

```diff
@@ -63,515 +65,506 @@ impl Default for PermissionState {
+pub struct NetPermission(pub String, pub Option<u16>);
+
+impl NetPermission {
+  fn new<T: AsRef<str>>(host: &&(T, Option<u16>)) -> Self {
+    NetPermission(host.0.as_ref().to_string(), host.1)
+  }
+
+  pub fn from_string(host: String) -> Self {
+    let url = url::Url::parse(&format!("http://{}", host)).unwrap();
+    let hostname = url.host_str().unwrap().to_string();
+
+    NetPermission(hostname, url.port())
+  }
+}
+
+impl fmt::Display for NetPermission {
+  fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+    f.write_str(&match self.1 {
+      None => self.0.clone(),
+      Some(port) => format!("{}:{}", self.0, port),
+    })
+  }
+}
+
+impl UnaryPermission<NetPermission> {
+  pub fn query<T: AsRef<str>>(
+    &self,
+    host: &Option<&(T, Option<u16>)>,
+  ) -> PermissionState {
+    if self.global_state == PermissionState::Denied
+      && match host.as_ref() {
+        None => true,
+        Some(host) => match host.1 {
+          None => self
             .denied_list
-            .retain(|path| !resolved_path.starts_with(path));
-          self.write.denied_list.insert(resolved_path);
-          self.write.global_state = PermissionState::Denied;
-          return PermissionState::Denied;
-        }
+            .iter()
+            .any(|host_| host.0.as_ref() == host_.0),
+          Some(_) => self.denied_list.contains(&NetPermission::new(host)),
+        },
       }
-      state
-    } else {
-      let state = self.query_write(&None);
-      if state == PermissionState::Prompt {
-        if permission_prompt("Deno requests write access") {
-          self.write.granted_list.clear();
-          self.write.global_state = PermissionState::Granted;
-          return PermissionState::Granted;
-        } else {
-          self.write.global_state = PermissionState::Denied;
-          return PermissionState::Denied;
+    {
+      PermissionState::Denied
+    } else if self.global_state == PermissionState::Granted
+      || match host.as_ref() {
+        None => false,
+        Some(host) => {
+          self.granted_list.contains(&NetPermission::new(&&(
+            host.0.as_ref().to_string(),
+            None,
+          )))
+            || self.granted_list.contains(&NetPermission::new(host))
         }
       }
-      state
+    {
+      PermissionState::Granted
+    } else {
+      PermissionState::Prompt
     }
   }
 
-  pub fn request_net<T: AsRef<str>>(
+  pub fn request<T: AsRef<str>>(
     &mut self,
     host: &Option<&(T, Option<u16>)>,
   ) -> PermissionState {
     if let Some(host) = host {
-      let state = self.query_net(&Some(host));
+      let state = self.query(&Some(host));
       if state == PermissionState::Prompt {
-        let host_string = format_host(host);
-        if permission_prompt(&format!(
-          "Deno requests network access to \"{}\"",
-          host_string,
-        )) {
+        let host = NetPermission::new(host);
```

Comment:
Nitpick: it's quite hard to read so many nested conditionals, maybe you could reorganize the code to return early instead?